sympy/polys/matrices/dense.py

```diff
@@ -195,6 +495,33 @@ def ddm_idet(a, K):
 
 
 def ddm_iinv(ainv, a, K):
+    """ainv  <--  inv(a)
+
+    Compute the inverse of a matrix $a$ over a field $K$ using Gauss-Jordan
+    elimination. The result is stored in $ainv$.
+
+    Uses division in the ground domain which should be an exact field.
+
+    Examples
+    ========
+
+    >>> from sympy.polys.matrices.ddm import ddm_iinv, ddm_imatmul
+    >>> from sympy import QQ
+    >>> a = [[QQ(1), QQ(2)], [QQ(3), QQ(4)]]
+    >>> ainv = [[None, None], [None, None]]
```

Comment:
```suggestion
    >>> ainv = []
```

Reply:
It is not really intended that that would be able to work...

Reply:
In retrospect it is probably a mistake that these operations work in place. You can see in this example that it does not give any performance benefit but it does complicate the interface.

I would rather not change that right now though.

Reply:
I don't understand how it complicates the interface. You need somewhere to put the result if you aren't going to return it. I am only suggesting that you change the docstring to show that you don't have to worry about what shape or what values must be in the `result` (mutable list) that you send: just send an empty list and it will be filled.

Reply:
> just send an empty list and it will be filled.

I don’t think that there can be a guarantee that that may work in the future though

Reply:
I think that the original intention was that `ainv` might be a `DDM` (which is a list subclass) and then this would be like:
```python
In [28]: from sympy.polys.matrices.ddm import DDM

In [29]: from sympy.polys.matrices.dense import ddm_iinv

In [30]: M = DDM([[QQ(1), QQ(2)], [QQ(3), QQ(4)]], (2, 2), QQ)

In [31]: Minv = DDM([[QQ(0), QQ(0)], [QQ(0), QQ(0)]], (2, 2), QQ)

In [32]: ddm_iinv(Minv, M, QQ)

In [33]: Minv
Out[33]: DDM([[mpq(-2,1), mpq(1,1)], [mpq(3,2), mpq(-1,2)]], (2, 2), QQ)

In [34]: type(Minv)
Out[34]: sympy.polys.matrices.ddm.DDM
```
In fact that is how it is used:
https://github.com/sympy/sympy/blob/5cc8f1dfe02f1bbea83ff08647541010a8279171/sympy/polys/matrices/ddm.py#L635-L638
I think the intention was that this function could be used with a list of lists but could also be used with `DDM` and would preserve the type either way.

If it is ever true that this function could changed to be more efficient like that then the future implementation would not be compatible with just passing in an empty list. For it to work with `DDM` we need the input `DDM` to have the correct shape and domain attributes.

Altogether I think it is just a mistake though and this should always have returned a new list. There might be extremely marginal performance benefit from this being in place but I doubt it. I think I had imagined that there would be lots of low-level code that makes many calls to functions like `ddm_imatmul` and `ddm_iinv` and could save performance by reusing allocated memory etc. Partly I wanted these functions to be designed so that they could be used without a domain e.g. for some type that supports the relevant operations like `+` and `*` but does not have a formal associated poly domain. I hoped that we could use that to replace the old ad-hoc matrices like `RawMatrix`, `NewMatrix`, `PolyMatrix` etc. In the end those were just all changed to use `DomainMatrix` though.

I see now that actually most of the more complicated routines probably only make sense to implement at the `DomainMatrix` level and we only really need the most primitive routines here at the `ddm_` level.

I also imagined that these functions might make sense for all the places in the codebase that use ad-hoc matrices e.g. like:
https://github.com/sympy/sympy/blob/5cc8f1dfe02f1bbea83ff08647541010a8279171/sympy/polys/fglmtools.py#L74-L84
I now just think that all of these things in the codebase should just use `DomainMatrix` and so there isn't really much benefit in the ``ddm_imatmul` etc routines for anything beyond making `DomainMatrix` work.

Even if we do want to work with list of lists then still it would be better just to have `ainv = ddm_inv(a, K)` that returns a new list.

Reply:
> Even if we do want to work with list of lists then still it would be better just to have ainv = ddm_inv(a, K) that returns a new list.

I generally advice to do static typing for these sort if possible, than having to be worrying about if you wrote something that consistently returns or not.

Reply:
I'm not going to change this right now but a future PR could add out of place versions for any current in place functions. Possibly the in place versions could be removed or otherwise the docstring could just be changed to call the out of place version of the function.

Reply:
Not sure if anyone will read this now that the PR has merged (and I might not even see the answer) but if you pass a list of lists as is done in the docstring, then it is pointless to fill it with anything. Neither the content nor the shape of the contents have any impact on what is returned -- and what is returned is a list of lists. All I was suggesting that be changed was the docstring to show a list instead of a list filled with lists filled with values (none of which matter and should not even be thought of when calling the routine that only returns a list of lists that replaces the contents of what was passed).

Reply:
Usually, the things to put on the docstrings should be the best practices, or the things that are invariant to the code

Reply:
But what is the best practice when the routine does not look at the contents or shape of what you are passing to it? In one docstring it is filled like this `ainv = [[None]*3 for i in range(3)]` and in another like `ainv = [[Q(1)]*3 for i in range(3)]`. None of this matters if the routine modifies like this `rv[:] = [row[:n] for row in smthng]`. We are making the user reading the docstring think about the contents of `ainv` when it doesn't matter.